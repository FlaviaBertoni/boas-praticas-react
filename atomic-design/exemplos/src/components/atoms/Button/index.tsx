import { ThemeColorsKeys } from '../../../styles/theme';
import { ContainedButton, OutlinedButton, TextButton } from './styles';

export type ButtonVariants = 'text' | 'outlined' | 'contained';
interface ButtonProps {
  color?: ThemeColorsKeys;
  variant?: ButtonVariants;
  width?: string;
  className?: string;
}

const Button: React.FC<ButtonProps> = ({ variant, color, width, className, children }) => {
  const buttonVariant = variant || 'outlined';
  return (
    <>
      {buttonVariant === 'contained' && (
        <ContainedButton color={color} width={width} className={className}>
          {children}
        </ContainedButton>
      )}
      {buttonVariant === 'outlined' && (
        <OutlinedButton color={color} width={width} className={className}>
          {children}
        </OutlinedButton>
      )}
      {buttonVariant === 'text' && (
        <TextButton color={color} width={width} className={className}>
          {children}
        </TextButton>
      )}
    </>
  );
};

export default Button;
