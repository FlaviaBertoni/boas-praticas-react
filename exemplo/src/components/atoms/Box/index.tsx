import { forwardRef } from 'react';
import { ThemeColorsType } from '../../../styles/theme';
import { Container } from './Box.styles';

export interface BoxProps extends React.HTMLAttributes<HTMLElement> {
  direction?: 'column' | 'row' | 'column-reverse' | 'row-reverse';
  alignItems?: 'normal' | 'stretch' | 'center' | 'start' | 'end' | 'baseline' | 'stretch';
  alignContent?: 'normal' | 'center' | 'space-between' | 'space-around' | 'space-evenly';
  justifyItems?: 'normal' | 'stretch' | 'center' | 'start' | 'end' | 'baseline';
  justifyContent?: 'normal' | 'center' | 'space-between' | 'space-around' | 'space-evenly';
  borderRadius?: string;
  bgColor?: keyof ThemeColorsType;
  padding?: string;
  margin?: string;
  border?: string;
  width?: string;
  maxWidth?: string;
  height?: string;
  as?: keyof JSX.IntrinsicElements;
  className?: string;
  ref?: React.ForwardedRef<any>;
  children?: React.ReactNode;
}

const Box = forwardRef(
  (
    {
      direction = 'row',
      alignItems = 'normal',
      alignContent = 'normal',
      justifyContent = 'normal',
      justifyItems = 'normal',
      borderRadius = '0',
      bgColor,
      padding = 'unset',
      margin = 'unset',
      border = 'none',
      width = 'auto',
      height = 'auto',
      maxWidth = 'unset',
      as,
      className,
      children,
      ...props
    }: BoxProps,
    ref: React.ForwardedRef<HTMLElement>,
  ) => (
    <Container
      direction={direction}
      borderRadius={borderRadius}
      bgColor={bgColor}
      padding={padding}
      margin={margin}
      border={border}
      width={width}
      maxWidth={maxWidth}
      height={height}
      alignItems={alignItems}
      alignContent={alignContent}
      justifyContent={justifyContent}
      justifyItems={justifyItems}
      as={as}
      className={className}
      ref={ref}
      {...props}
    >
      {children}
    </Container>
  ),
);

export default Box;
