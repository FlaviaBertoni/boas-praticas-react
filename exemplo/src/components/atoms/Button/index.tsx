import { ThemeColorsKeys } from '../../../styles/theme';
import { ContainedButton, OutlinedButton, TextButton } from './styles';

export type ButtonVariants = 'text' | 'outlined' | 'contained';
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  color?: ThemeColorsKeys;
  variant?: ButtonVariants;
  width?: string;
  margin?: string;
  className?: string;
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

const Button: React.FC<ButtonProps> = ({ variant, color, width, className, children, margin, onClick, ...rest }) => {
  const buttonVariant = variant || 'outlined';
  return (
    <>
      {buttonVariant === 'contained' && (
        <ContainedButton onClick={onClick} color={color} width={width} margin={margin} className={className} {...rest}>
          {children}
        </ContainedButton>
      )}
      {buttonVariant === 'outlined' && (
        <OutlinedButton onClick={onClick} color={color} width={width} margin={margin} className={className} {...rest}>
          {children}
        </OutlinedButton>
      )}
      {buttonVariant === 'text' && (
        <TextButton onClick={onClick} color={color} width={width} margin={margin} className={className} {...rest}>
          {children}
        </TextButton>
      )}
    </>
  );
};

export default Button;
