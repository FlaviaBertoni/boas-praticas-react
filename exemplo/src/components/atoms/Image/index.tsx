import { Container } from './Image.style';

export interface ImageProps {
  width?: string;
  height?: string;
  margin?: string;
  padding?: string;
  borderRadius?: string;
  objectFit?: 'cover' | 'contain' | 'fill' | 'unset';
  imgPath: string;
  alt: string;
  className?: string;
  as?: keyof JSX.IntrinsicElements;
}

const Image = ({
  imgPath,
  alt,
  width = '100%',
  height = '100%',
  objectFit = 'contain',
  borderRadius = '0',
  margin = '0',
  padding = '0',
  className,
  as,
}: ImageProps): JSX.Element => (
  <Container
    borderRadius={borderRadius}
    width={width}
    height={height}
    objectFit={objectFit}
    margin={margin}
    padding={padding}
    className={className}
    as={as}
  >
    <img src={imgPath} alt={alt} />
  </Container>
);

export default Image;
