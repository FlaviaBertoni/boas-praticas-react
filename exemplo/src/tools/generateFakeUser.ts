const generateFakeUser = () => ({
  name: 'Fake test user',
  email: 'fakeuser@fake.dev',
  password: 'F@k3p4ssw0rd',
});

export default generateFakeUser;
