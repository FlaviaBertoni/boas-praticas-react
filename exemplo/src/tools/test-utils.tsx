import { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import theme from '../styles/theme';

const AllTheProviders: FC = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'queries'>) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
