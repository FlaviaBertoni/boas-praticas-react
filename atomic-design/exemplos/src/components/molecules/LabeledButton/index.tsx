import React from 'react';
import theme, { ThemeColorsKeys } from '../../../styles/theme';
import Button, { ButtonVariants } from '../../atoms/Button';
import Label, { labelSize } from '../../atoms/Label';

export type LabeledButtonSizes = 'small' | 'large';

interface LabeledButtonProps {
  size: LabeledButtonSizes;
  variant: ButtonVariants;
  color?: ThemeColorsKeys;
  className?: string;
}

const LabeledButton: React.FC<LabeledButtonProps> = ({ children, size, variant, color, className }) => {
  const buttonWidth = size === 'large' ? '224px' : '112px';
  const labelSize: labelSize = size === 'large' ? 'medium' : 'small';
  const labelColor: ThemeColorsKeys | undefined =
    variant === 'contained' ? (theme.isDarkColor(color) ? 'white' : 'black') : color;

  return (
    <Button width={buttonWidth} variant={variant} color={color} className={className}>
      <Label size={labelSize} weight="bold" color={labelColor}>
        {children}
      </Label>
    </Button>
  );
};

export default LabeledButton;
