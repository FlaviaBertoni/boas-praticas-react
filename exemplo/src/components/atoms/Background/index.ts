import styled from 'styled-components';
import Box from '../Box';

interface BackgroundProps {
  url: string;
}

export const Background = styled(Box)<BackgroundProps>`
  flex: 1;
  background: ${({ url }) => `url(${url}) no-repeat center`};
  background-size: cover;
`;
