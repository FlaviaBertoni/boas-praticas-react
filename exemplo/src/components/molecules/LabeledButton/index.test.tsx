import { render } from '../../../tools/test-utils';
import LabeledButton, { LabeledButtonSizes } from '.';
import theme, { ThemeColorsType } from '../../../styles/theme';
import { ButtonVariants } from '../../atoms/Button';

const colors = Object.keys(theme.colors);
const variants: ButtonVariants[] = ['text', 'outlined', 'contained'];
const sizes: LabeledButtonSizes[] = ['small', 'large'];

const matriz: Array<Array<String>> = [];

/*
 * Just for test purposes.
 * Only will work ordered with the biggest to the smallest array.
 */
colors.map(color => sizes.map(size => variants.map(variant => matriz.push([size, variant, color]))));

describe('Test case for LabeledButton component', () => {
  // We can even test a list of lists using each:
  it.each(matriz)(
    'should be able to render the LabeledButton component with %s size, %s variant, %s color and match with the snapshot',
    (size, color, variant) => {
      const { container } = render(
        <LabeledButton
          size={size as LabeledButtonSizes}
          color={color as keyof ThemeColorsType}
          variant={variant as ButtonVariants}
        >
          unit test
        </LabeledButton>,
      );

      expect(container).toMatchSnapshot();
    },
  );
});
