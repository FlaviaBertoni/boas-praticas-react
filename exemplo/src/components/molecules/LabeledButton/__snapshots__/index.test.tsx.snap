// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, contained variant, black color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, contained variant, card-bg color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, contained variant, dark color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, contained variant, disable color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, contained variant, primary color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, contained variant, transparent color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, contained variant, white color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, outlined variant, black color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, outlined variant, card-bg color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, outlined variant, dark color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, outlined variant, disable color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, outlined variant, primary color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, outlined variant, transparent color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, outlined variant, white color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, text variant, black color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, text variant, card-bg color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, text variant, dark color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, text variant, disable color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, text variant, primary color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, text variant, transparent color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with large size, text variant, white color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, contained variant, black color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, contained variant, card-bg color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, contained variant, dark color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, contained variant, disable color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, contained variant, primary color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, contained variant, transparent color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, contained variant, white color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, outlined variant, black color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, outlined variant, card-bg color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, outlined variant, dark color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, outlined variant, disable color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, outlined variant, primary color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, outlined variant, transparent color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, outlined variant, white color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, text variant, black color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, text variant, card-bg color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, text variant, dark color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, text variant, disable color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, text variant, primary color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, text variant, transparent color and match with the snapshot 1`] = `<div />`;

exports[`Test case for LabeledButton component should be able to render the LabeledButton component with small size, text variant, white color and match with the snapshot 1`] = `<div />`;
